src/pudl/helpers.py

```diff
@@ -339,6 +339,107 @@ def is_doi(doi):
+    columns as well as additional columns specified by the ``on_cols`` argument.
+
+    Args:
+        left: The higher frequency "data" dataframe. Typically monthly in our use
+            cases. E.g. ``generation_eia923``. Must contain columns specified by
+            ``left_date_col`` and ``on_cols`` argument.
+        right: The lower frequency "attribute" dataframe. Typically annual in our uses
+            cases. E.g. ``generators_eia860``. Must contain columns specified by
+            ``right_date_col`` and ``on_cols`` argument.
+        left_date_col: Column in ``left`` containing datetime like data. Default is
+            ``report_date``. Must be convertible to a Datetime using
+            :func:`pandas.to_datetime`
+        right_date_col: Column in ``right`` containing datetime like data. Default is
+            ``report_date``. Must be convertible to a Datetime using
+            :func:`pandas.to_datetime`
+        on_cols: Columns to merge on in addition to newly created temporal columns. Typically
+            ID columns like ``plant_id_eia``, ``generator_id`` or ``boiler_id``.
+        gran: The desired granularity that the less granular dataframe should be
+            broadcast to. For now, this is the granularity of the more granular
+            dataframe.
+        report_time_in_period: When data is reported in the less granular dataframe.
+            When equal to "first" this means data in the right dataframe is reported
+            on the first date in the time period. E.g. annual dataframes have January
+            1st as the date. This is what happens by defualt if only a year or year-month
+            are provided to :func:`pandas.to_datetime` as strings.
+        merge_type: How the dataframes should be merged.
+
+
+    Returns:
+        Merged contents of left and right input dataframes.
+
+    Raises:
+        ValueError: if ``left_date_col`` or ``right_date_col`` columns are missing from their
+            respective input dataframes.
+        ValueError: if any of the labels referenced in ``on_cols`` are missing from either
+            the left or right dataframes.
+
+    """
+    # function breaks up a datetime column into separate columns
+    def assign_date_cols(df, date_col_name):
+        df[f"{date_col_name}"] = pd.to_datetime(df[f"{date_col_name}"])
+        df = df.assign(
+            year=lambda x: x[f"{date_col_name}"].dt.year,
+            quarter=lambda x: x[f"{date_col_name}"].dt.quarter,
+            month=lambda x: x[f"{date_col_name}"].dt.month,
+            day=lambda x: x[f"{date_col_name}"].dt.day,
+        )
+        return df
+
+    # add rows with an extra year for each unique group of rows, will be dropped
+    # this fixes an off by one issue with upsampling
+    if report_time_in_period == "first":
+        extra_year_df = (
+            right.groupby(on_cols)[right_date_col].max() + pd.offsets.YearBegin()
```

Comment:
I need to change this so the offset (`pd.offsets.YearBegin()` for now) change based on the granularity of the right dataframe. Maybe I'll check what the offset is between two rows and use this time delta as the offset.